test/grisp_connect_test_server.erl

```diff
@@ -198,8 +199,20 @@ wait_disconnection() ->
 
 %--- Websocket Callbacks -------------------------------------------------------
 
-init(Req, State) ->
-    {cowboy_websocket, Req, State}.
+init(Req, Opts) ->
+    ExpVer = maps:get(expected_protocol, Opts, <<"grisp-io-v1">>),
+    SelVer = maps:get(selected_protocol, Opts, <<"grisp-io-v1">>),
+    case cowboy_req:header(<<"sec-websocket-protocol">>, Req) of
+        undefined ->
+            Req2 = cowboy_req:reply(400, #{}, <<"No protocol specified">>, Req),
+            {ok, Req2, Opts};
+        ExpVer ->
+            Req2 = cowboy_req:set_resp_header(<<"sec-websocket-protocol">>, SelVer, Req),
+            {cowboy_websocket, Req2, Opts};
+        _ ->
+            Req2 = cowboy_req:reply(400, #{}, <<"Unsupported Protocol">>, Req),
```

Comment:
https://datatracker.ietf.org/doc/html/rfc6455#section-4.2.2
> If the client's handshake did not contain such a header field or if the server does not agree to any of the client's requested subprotocols, the only acceptable value is null.  The absence of such a field is equivalent to the null value (meaning that if the server does not wish to agree to one of the suggested subprotocols, it MUST NOT send back a |Sec-WebSocket-Protocol| header field in its response).

The correct response here would be to not send the header back or set the header to `null`.

Reply:
As I understand it, this only means that IF the server upgrade the connection, the header should either not be specified or `null`, but that doesn't mean the server could not decide to NOT upgrade the connection because it doesn't make sense for the server to continue if the client protocol is not accepted.... By continuing without the header or setting it to `null`, the server would rely on the client to validate the protocol handshake, when it already knows it is not correct, if the client just ignore the header, the handshake would be in an unstable state.

Reply:
Ok we keep it with status 400 then.